Add D-FINE Model into Transformers

VladOS95-cyber opened the pull request:
# What does this PR do?

D-FINE, a powerful real-time object detector that achieves outstanding localization precision by redefining the bounding box regression task in DETR models. D-FINE comprises two key components: Fine-grained Distribution Refinement (FDR) and Global Optimal Localization Self-Distillation (GO-LSD).

As backbone, it is used HGNet-V2 (High Performance GPU Net) image classification model.

This PR add D-FINE and HGNet-V2 into the Transformers library. There is a new thing in transformers called modular, which adds new models by creating a modeling_modelname.py file. Since D-FINE updates several RT-DETR arch parts while keeping the rest of the model unchanged, it serves as an ideal use case for this modular approach.

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. https://github.com/huggingface/transformers/issues/35283
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?


## Who can review?

@qubvel @Rocketknight1 @ArthurZucker @NielsRogge 

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker
- vision models: @amyeroberts, @qubvel
- speech models: @ylacombe, @eustlb
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @zucchini-nlp (visual-language models) or @gante (all others)
- pipelines: @Rocketknight1
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @SunMarc
- chat templates: @Rocketknight1

Integrations:

- deepspeed: HF Trainer/Accelerate: @muellerzr
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc @MekkCyber

Documentation: @stevhliu

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

VladOS95-cyber committed:
copy the last changes from broken PR

VladOS95-cyber commented:
Hi @qubvel! I decided to create a new PR as previous one (https://github.com/huggingface/transformers/pull/35400) was broken, sorry for the mess. In this PR all latest changes

VladOS95-cyber committed:
small format

Cyrilvallez commented:
Hey very sorry, I had very low bandwith last week as I was presenting at a conference! Are you still experiencing issue with how modular works, or is it resolved?

VladOS95-cyber commented:
> Hey very sorry, I had very low bandwith last week as I was presenting at a conference! Are you still experiencing issue with how modular works, or is it resolved?

Hey @Cyrilvallez! Thank you for your answer, yes, there are still some issues: 1. Wrong namespace import for modeling_d_fine_resnet.py where conversion script provides from .configuration_d_fine_resnet_resnet. 2. DFine config conversion using modular does not work properly because it provides a call super().init(is_encoder_decoder=is_encoder_decoder, **kwargs) too eraly in the very beginning of init and before all self.args assignments, which is wrong as well.

Cyrilvallez commented:
Ok @VladOS95-cyber, I looked into it 🤗
1. is indeed an issue, as you have 2 modular files in the same folder, this was unprecedented and we were not parsing correctly: https://github.com/huggingface/transformers/pull/36287 will solve it
2. This is actually by design, as usually (but not always, true) `super().__init__()` calls are placed at the beginning. In this instance it should not be an issue, is it? Meaning that I don't think the order has any importance in this case

Cyrilvallez commented:
Oh, or is 2. an issue because of the `attribute_map`? I may have seen this issue when reviewing RtDetrV2. But in this case, I would argue that using the `attribute_map` itself is usually very confusing, and it would definitely be best not to if possible (e.g. by renaming the arg directly)

VladOS95-cyber commented:
> Ok @VladOS95-cyber, I looked into it, thanks for the feedback 🤗
> 
> 1. is indeed an issue, as you have 2 modular files in the same folder, this was unprecedented and we were not parsing correctly: [[modular] allow multiple modular files in the same model folder #36287](https://github.com/huggingface/transformers/pull/36287) will solve it
> 2. This is actually by design, as usually (but not always, true) `super().__init__()` calls are placed at the beginning. In this instance it should not be an issue, is it? Meaning that I don't think the order has any importance in this case

Hi @Cyrilvallez! Well, I suppose it is an issue in case of config and attribute map uage, yes. Because if we call super().__init__(kwargs**) in the beginning and do args.self assignment later, we override some attributes values that were passed before.  In RTDetrV2 this problem was resolved by implementing config from the sctratch, but in our case, we reuse RTDetrConfig, so this problem came up unfortunately.

VladOS95-cyber commented:
@Cyrilvallez So, what do you suggest in thi case?

qubvel commented on src/transformers/models/d_fine/modular_d_fine_resnet.py:
```diff
@@ -0,0 +1,414 @@
+    def __init__(
+        self,
+        stem_channels=[3, 32, 48],
+        stage_config=[
+            [48, 48, 128, 1, False, False, 3, 6],
+            [128, 96, 512, 1, True, False, 3, 6],
+            [512, 192, 1024, 3, True, True, 5, 6],
+            [1024, 384, 2048, 1, True, True, 5, 6],
+        ],
```
same here

qubvel commented on src/transformers/models/d_fine/modular_d_fine_resnet.py:
```diff
@@ -0,0 +1,414 @@
+            self.layers.append(block)
+
+        # feature aggregation
+        total_chs = in_channels + layer_num * middle_channels
```
```suggestion
        total_channels = in_channels + layer_num * middle_channels
```

qubvel commented on src/transformers/models/d_fine/modular_d_fine_resnet.py:
```diff
@@ -0,0 +1,414 @@
+
+
+class DFineResNetStage(nn.Module):
+    def __init__(
+        self,
+        in_channels: int,
+        mid_channels: int,
+        out_channels: int,
+        block_num: int,
+        layer_num: int,
+        downsample: bool = True,
+        light_block: bool = False,
+        kernel_size: int = 3,
+        aggregation: str = "se",
+        drop_path: float = 0.0,
+        use_learnable_affine_block: bool = False,
+    ):
```
btw can we initialize it from config with the given stage index as an additional argument?

qubvel commented on tests/models/d_fine/test_modeling_d_fine.py:
```diff
@@ -0,0 +1,800 @@
+    test_torchscript = False
+    test_pruning = False
+    test_head_masking = False
+    test_missing_keys = False
```
```suggestion
    test_missing_keys = False
    test_torch_exportable = True
```

qubvel reviewed (commented):
Thanks, @VladOS95-cyber for addressing the comments, refactoring, and adding type hints! I left a few minor comments, but it's in good shape. As the next step, let's also ensure:

 - Modular issues are resolved (cc @Cyrilvallez)
 - `torch.compile` (fullgraph) and `torch.export` are supported (see comments)
 - All checkpoints can be converted (please provide links for converted checkpoints on the Hub) and logits/boxes match the original implementation
 - Model can be fine-tuned ([here](https://github.com/qubvel/transformers-notebooks/blob/main/notebooks/RT_DETR_v2_finetune_on_a_custom_dataset.ipynb) is an example for RT-DETRv2, it would work with the change of checkpoint path only)
 - CI is happy: consistency, docs, style issues are resolved
 
Thanks for the great work!

VladOS95-cyber commented on src/transformers/models/d_fine/configuration_d_fine_resnet.py:
```diff
@@ -0,0 +1,94 @@
+        self.hidden_act = hidden_act
+        self.downsample_in_first_stage = downsample_in_first_stage
+        self.downsample_in_bottleneck = downsample_in_bottleneck
+        self.stage_names = ["stem"] + [f"stage{idx}" for idx in range(1, len(depths) + 1)]
```
you are right, no need

VladOS95-cyber commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -0,0 +1,830 @@
+
+        self.ms_deformable_attn_core = multi_scale_deformable_attention_v2
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        attention_mask: Optional[torch.Tensor] = None,
+        reference_points=None,
+        encoder_hidden_states=None,
+        spatial_shapes=None,
+    ) -> tuple[torch.Tensor, torch.Tensor]:
```
Hi Pavel, are you sure that torch.compile works as expected for RtDetrV2? I encounter some issues with that not only with my implementation but I checked RtDetrV2 behaviour and I cannot execute smth like that: `compiled_model = torch.compile(model, fullgraph=True)
output = compiled_model(inputs)`. Compilation is executed successfully, yes, but forward prop does not work not for my model nor for RtDetrV2.

VladOS95-cyber replied:
Maye I do somethid wrong, but I just compile and then try to execute prediction on input. I got smth like that: 
![Снимок](https://github.com/user-attachments/assets/f87d0f73-a3ff-4b7d-a2bc-2fb5b6388c39)

VladOS95-cyber commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -0,0 +1,830 @@
+    """
+
+    def __init__(self, config: DFineConfig):
+        self.eval_idx = config.eval_idx if config.eval_idx >= 0 else config.decoder_layers + config.eval_idx
+        super().__init__(config=config)
```
If we put `self.eval_idx = config.eval_idx if config.eval_idx >= 0 else config.decoder_layers + config.eval_idx after super().__init__(config=config)`, then ```
self.layers = nn.ModuleList(
            [DFineDecoderLayer(config=config) for _ in range(config.decoder_layers)]
            + [DFineDecoderLayer(config=config) for _ in range(config.decoder_layers - self.eval_idx - 1)]
        )
``` is initialized before we assign self.eval_idx and exception will be thrown, I suppose it is behaviour of modular conversion script

VladOS95-cyber committed:
some fixes and refactoring after review

VladOS95-cyber committed:
format

VladOS95-cyber committed:
add config attr for loss

VladOS95-cyber commented:
Hi @qubvel! There is a short summary what we've achieved so far by this moment:
1. Modular issues are resolved (cc @Cyrilvallez) - still in progress. Moreover, I would like to discuss what we are going to do with config and attribute map problem.
2. `torch.compile (fullgraph)` works and compiles the model, but I cannot execute this model later on inference as it is failed with exception I provided in comments. But I checked the behavior of other models in this case, it is the same. So, I assume either I do something wrong or it is expected. 
3. `torch.export` is supported. I added it into tests and it was executed successfully. 
4. All main checkpoints were converted: https://huggingface.co/vladislavbro/dfine_x_coco, https://huggingface.co/vladislavbro/dfine_l_coco, https://huggingface.co/vladislavbro/dfine_m_coco, https://huggingface.co/vladislavbro/dfine_s_coco. All logits/boxes are matched with original implementation
5. Model is able to be fine tuned, I checked on the script you provided and changed checkpoint on d-fine one, and it worked. 
6. CI is not happy yet because it depends on issues from modular conversion. Once it will be resolved, CI will be completed.

qubvel reviewed (commented):
Hi @VladOS95-cyber, thanks for iterating and addressing comments! I left a few more comments, but in general looks great!

I am having difficulty running fine-tuning for the "m" checkpoint, while the "l" checkpoint is fine. Can you please double-check the modeling code, losses and initialization strategy? Here is the error I got:

```
[/usr/local/lib/python3.11/dist-packages/transformers/loss/loss_for_object_detection.py](https://localhost:8080/#) in generalized_box_iou(boxes1, boxes2)
    416     # so do an early check
    417     if not (boxes1[:, 2:] >= boxes1[:, :2]).all():
--> 418         raise ValueError(f"boxes1 must be in [x0, y0, x1, y1] (corner) format, but got {boxes1}")
    419     if not (boxes2[:, 2:] >= boxes2[:, :2]).all():
    420         raise ValueError(f"boxes2 must be in [x0, y0, x1, y1] (corner) format, but got {boxes2}")

ValueError: boxes1 must be in [x0, y0, x1, y1] (corner) format, but got tensor([[4.0848e-02, 6.0155e-02, 9.7476e-01, 9.4098e-01],
        [4.3698e-02, 1.0271e-01, 9.6435e-01, 8.6763e-01],
        [4.0578e-06, 6.4026e-06, 1.4135e-05, 2.2186e-05],
        ...,
        [9.9940e-01, 3.7292e-02, 1.0063e+00, 4.3024e-02],
        [9.8414e-01, 4.3545e-02, 1.0219e+00, 4.8508e-02],
        [7.8416e-01, 6.7056e-01, 8.3837e-01, 7.0462e-01]], device='cuda:0')
```

P.S. `torch.compile(model, fullgraph=True)` works fine with my env, thanks!

VladOS95-cyber commented on src/transformers/models/d_fine/modular_d_fine_resnet.py:
```diff
@@ -0,0 +1,476 @@
+        use_learnable_affine_block (`bool`, *optional*, defaults to False):
+            Whether to use Learnable Affine Blocks (LAB) in the network.
+            LAB adds learnable scale and bias parameters after certain operations.
+        aggregation (`str`, *optional*, defaults to se):
```
The other option is "ese". nad there is a specific module to handle it called DFineResNetEseModule

VladOS95-cyber replied:
I think we can remove it

VladOS95-cyber commented on src/transformers/models/d_fine/modeling_d_fine_resnet.py:
```diff
@@ -0,0 +1,494 @@
+                    kernel_size=kernel_size,
+                    light_block=light_block,
+                    aggregation=aggregation,
+                    drop_path=drop_path[i] if isinstance(drop_path, (list, tuple)) else drop_path,
```
Yes, you are right

VladOS95-cyber commented on src/transformers/models/d_fine/modeling_d_fine_resnet.py:
```diff
@@ -0,0 +1,494 @@
+                aggregation_squeeze_conv,
+                aggregation_excitation_conv,
+            )
+        else:
+            aggregation_conv = DFineResNetConvLayer(
+                total_channels,
+                out_channels,
+                kernel_size=1,
+                stride=1,
+                use_learnable_affine_block=use_learnable_affine_block,
+            )
+            ese_aggregation_module = DFineResNetEseModule(out_channels)
+            self.aggregation = nn.Sequential(
+                aggregation_conv,
+                ese_aggregation_module,
+            )
+        self.drop_path = nn.Dropout(drop_path) if drop_path else nn.Identity()
```
I checked original implementation, and it seems so, yes, there is "ese" option and related module but I really did not find where it could be setup

qubvel commented on src/transformers/models/d_fine/modeling_d_fine.py:
```diff
@@ -0,0 +1,2266 @@
+        >>> url = "http://images.cocodataset.org/val2017/000000039769.jpg"
+        >>> image = Image.open(requests.get(url, stream=True).raw)
+
+        >>> image_processor = DFineImageProcessor.from_pretrained("PekingU/DFine_r50vd")
+        >>> model = DFineForObjectDetection.from_pretrained("PekingU/DFine_r50vd")
```
I suppose smth like

```python
def forward(**super_kwargs):
    """
    ```
    <updated python example goes here>
    ```
    """
    super().forward()
```

should work for modular, otherwise we can replace the whole docstring

VladOS95-cyber committed:
some fixes and refactoring

VladOS95-cyber committed:
fix copies

VladOS95-cyber committed:
fix style

qubvel commented:
@VladOS95-cyber hmm, yes, there ware some PRs regarding `from_pretrained` refactoring.. it might be broken, let me check

qubvel commented:
PR with fix merged 👍 thanks for reporting

VladOS95-cyber committed:
add test for d-fine resnet

VladOS95-cyber committed:
fix decoder layer prop

VladOS95-cyber committed:
fix dummies

VladOS95-cyber committed:
format init

VladOS95-cyber commented:
Hi @qubvel! Latest state: I added tests for resnet. I don't know how much time I checked modeling code but I could not find any problems that maybe a root cause of this issue with fine-tuning. I made small fix but it did not help and it seems weird as loss calculation is always failed on different steps and sometimes is not failed at all. Returned scores and boxes are the same as in original implementation. There are some differenses in states, yes, but there is logical explanation for that and it should not be a problem. Unfortunately, I am a little bit stacked right now...

VladOS95-cyber committed:
remove extra print

qubvel commented:
Hey @VladOS95-cyber, thanks for checking. I suppose it's not a blocker, and maybe we can figure it out later. At least inference works for all models, and fine-tuning for bigger models also works fine. I will take a deep dive next week and try to resolve the remaining issues to get this merged.

VladOS95-cyber commented:
Hey @qubvel! I am just curious, I found an issue https://github.com/huggingface/transformers/issues/36516. Could it be the case that an existing issue with fine-tuning for D-FINE is related to this one?

qubvel commented:
Well, it might be an issue, however, other models fine-tuning is OK with this dataset. This model might be a bit more sensitive to noisy labels, let me try to run with a clean one

VladOS95-cyber commented:
Hi @qubvel! I can resolve all CI tests problems, but to do so I need:
1. Move Resnet implementation into separate folder, because as I understood it is not going to be resolved according to Arthur comments in dedicated PR https://github.com/huggingface/transformers/pull/36287. 
2. We have to remove dependancy from RTDetrConfig and RtDetrResnetConfig in DFineConfig and DFineResnetConfig because we cannot pass some common attribute config unit tests and repository consistency tests as some args we do not use from RtDetr but we cannot remove them from init and repository tests consider them as not used anywhere and failed, this problem is also modular specific. 

So, what would you suggest to do with that?

qubvel commented:
Hey @VladOS95-cyber 

> Move Resnet implementation into separate folder, because as I understood it is not going to be resolved according to Arthur comments in dedicated PR https://github.com/huggingface/transformers/pull/36287.

Making two separate folders is OK, I would even say it's better to have them under the same folder.

> We have to remove dependancy from RTDetrConfig and RtDetrResnetConfig in DFineConfig and DFineResnetConfig because we cannot pass some common attribute config unit tests and repository consistency tests as there are some args we do not use from RtDetr but we cannot remove them from init and repository tests consider them as not used anywhere and failed, this problem is also modular specific.

I suppose we can use `del some_arg` in the DFine config, did you tried this approach? I case it does not help, I'm ok with removing dependency (+ adding a clear comment why we avoid inheritance, TODO statemnt, and link to PR comment, so we later can resolve it)

qubvel commented:
Ok, let's move forward with it's own config for DFine 🤗 

btw, tried fine-tuning for "S" model on clean CPPE-5 dataset - but still getting an error, debugging it right now

VladOS95-cyber commented:
> Ok, let's move forward with it's own config for DFine 🤗
> 
> btw, tried fine-tuning for "S" model on clean CPPE-5 dataset - but still getting an error, debugging it right now

Just one thing I forgot to mention. Last time when I executed fine tuning on L and M successfully was a testing with generate_anchors from original implementation from DFine repo. It is done slightly differently there and I even did it in a way to not calculate it on the fly but to reuse it from original saved weights. I tried to achieve full similarity in hidden_states and it is only possible when we explicitly download anchors from saved weights. I honestly don't know if it influences on current issue but I see any relations here actually.

qubvel commented:
hmm, interesting, from a quick look anchor generation is the same
https://github.com/Peterande/D-FINE/blob/2ef2a665d764604befe992af01ae9c3070103221/src/zoo/dfine/dfine_decoder.py#L731

Can you provide some details on the difference?

VladOS95-cyber commented:
> hmm, interesting, from a quick look anchor generation is the same https://github.com/Peterande/D-FINE/blob/2ef2a665d764604befe992af01ae9c3070103221/src/zoo/dfine/dfine_decoder.py#L731
> 
> Can you provide some details on the difference?

Yes, it is more or less the same, except, original implementation uses `anchors = torch.where(valid_mask, anchors, torch.inf)` in our case we use `anchors = torch.where(valid_mask, anchors, torch.tensor(torch.finfo(dtype).max, dtype=dtype, device=device))`. But it think it is not the problem, the problem is that ```
# init encoder output anchors and valid_mask
        if self.eval_spatial_size:
            anchors, valid_mask = self._generate_anchors()
            self.register_buffer("anchors", anchors)
            self.register_buffer("valid_mask", valid_mask)
        # init encoder output anchors and valid_mask
        if self.eval_spatial_size:
            self.anchors, self.valid_mask = self._generate_anchors()
``` and later it is recalculated for decoder input for example in `_get_decoder_input`. And when I try to compare downloaded anchors with computed ones it is completely different, which is fair enough because it seems that original code uses another logic to recalculate anchors later, like this: ```
def _get_decoder_input(
        self, memory: torch.Tensor, spatial_shapes, denoising_logits=None, denoising_bbox_unact=None
    ):
        # prepare input for decoder
        if self.training or self.eval_spatial_size is None:
            anchors, valid_mask = self._generate_anchors(spatial_shapes, device=memory.device)
        else:
            anchors = self.anchors
            valid_mask = self.valid_mask
        if memory.shape[0] > 1:
            anchors = anchors.repeat(memory.shape[0], 1, 1)

        # memory = torch.where(valid_mask, memory, 0)
        # TODO fix type error for onnx export
        memory = valid_mask.to(memory.dtype) * memory

        output_memory: torch.Tensor = self.enc_output(memory)
        enc_outputs_logits: torch.Tensor = self.enc_score_head(output_memory)

        enc_topk_bboxes_list, enc_topk_logits_list = [], []
        enc_topk_memory, enc_topk_logits, enc_topk_anchors = self._select_topk(
            output_memory, enc_outputs_logits, anchors, self.num_queries
        )
```

qubvel commented:
hmm, still not sure I got it, the original implementation using the same `self._generate_anchors` method either in `__init__` or in `forward` and the generated anchors should be the same in both cases, unless spatial shapes are different, right?

VladOS95-cyber commented:
> if memory.shape[0] > 1:
>             anchors = anchors.repeat(memory.shape[0], 1, 1)

Yes, sounds good, I sent to you via LinkedIn

VladOS95-cyber committed:
refactor modeling, move resnet into separate folder

VladOS95-cyber committed:
fix resnet config

VladOS95-cyber committed:
change resnet on hgnet_v2, add clamp into decoder

VladOS95-cyber committed:
fix init

VladOS95-cyber committed:
fix config doc

VladOS95-cyber committed:
fix init

VladOS95-cyber committed:
fix dummies

VladOS95-cyber committed:
fix config docs

VladOS95-cyber committed:
fix hgnet_v2 config typo

VladOS95-cyber committed:
format modular

VladOS95-cyber committed:
add image classification for hgnet, some refactoring

VladOS95-cyber committed:
format tests

VladOS95-cyber committed:
fix dummies

VladOS95-cyber committed:
fix init

VladOS95-cyber committed:
fix style

VladOS95-cyber committed:
fix init for hgnet v2

VladOS95-cyber committed:
fix index.md, add init rnage for hgnet

VladOS95-cyber committed:
fix conversion

VladOS95-cyber committed:
add missing attr to encoder

VladOS95-cyber committed:
add loss for d-fine, add additional output for rt-detr decoder

VladOS95-cyber committed:
tests and docs fixes

VladOS95-cyber committed:
fix rt_detr v2 conversion

VladOS95-cyber committed:
some fixes for loos and decoder output

VladOS95-cyber committed:
some fixes for loss

VladOS95-cyber committed:
small fix for converted modeling

VladOS95-cyber committed:
add n model config, some todo comments for modular

qubvel reviewed (commented):
Nice, HUGE thanks for working on the model loss, that's not an easy task!
Just a few nits and we are good to go!

The next step would be to convert all checkpoints (including objects365) and refine model cards!

VladOS95-cyber commented on src/transformers/loss/loss_d_fine.py:
```diff
@@ -0,0 +1,385 @@
+                    ).sum(-1)
+                )
+
+                batch_scale = 8 / outputs["pred_boxes"].shape[0]  # Avoid the influence of batch size per GPU
```
should we just provide outputs["pred_boxes"].shape[0] or no batch_scale at all?

VladOS95-cyber commented on src/transformers/loss/loss_rt_detr.py:
```diff
@@ -443,6 +443,8 @@ def RTDetrForObjectDetectionLoss(
     enc_topk_logits=None,
     enc_topk_bboxes=None,
     denoising_meta_values=None,
+    predicted_corners=None,
+    initial_reference_points=None,
```
yes, agree

VladOS95-cyber commented on src/transformers/models/rt_detr/modeling_rt_detr.py:
```diff
@@ -1466,9 +1482,11 @@ def forward(
+                new_reference_points = F.sigmoid(prediced_corners + inverse_sigmoid(reference_points))
                 reference_points = new_reference_points.detach()
+                intermediate_predicted_corners += (prediced_corners,)
+                initial_reference_points += (reference_points,)
```
yes, valid point, missed that

VladOS95-cyber replied:
We should provide just None then? Because the output structure must be the same as for D-FINE it is important

VladOS95-cyber committed:
convert script adjustments and fixes, small refact

qubvel commented:
Hey @Peterande, we are adding your awesome D-Fine model to Transformers! If you have time to review this PR, it would be super helpful to ensure we didn’t miss anything. Either way, huge thanks for your work, it’s truly appreciated!

VladOS95-cyber committed:
remove extra output for rt_detr

VladOS95-cyber committed:
make some outputs optionsl, fix conversion

qubvel commented on src/transformers/models/rt_detr/modeling_rt_detr.py:
```diff
@@ -137,10 +141,12 @@ class RTDetrDecoderOutput(ModelOutput):
-    intermediate_hidden_states: Optional[torch.FloatTensor] = None
-    intermediate_logits: Optional[torch.FloatTensor] = None
-    intermediate_reference_points: Optional[torch.FloatTensor] = None
+    last_hidden_state: torch.FloatTensor = None
+    intermediate_hidden_states: torch.FloatTensor = None
+    intermediate_logits: torch.FloatTensor = None
+    intermediate_reference_points: torch.FloatTensor = None
+    intermediate_predicted_corners: torch.FloatTensor = None
+    initial_reference_points: torch.FloatTensor = None
```
Please, make sure to add Optional[]

VladOS95-cyber replied:
right, missed that, fixed

qubvel commented on src/transformers/models/rt_detr/modeling_rt_detr.py:
```diff
@@ -200,10 +206,12 @@ class RTDetrModelOutput(ModelOutput):
-    intermediate_hidden_states: Optional[torch.FloatTensor] = None
-    intermediate_logits: Optional[torch.FloatTensor] = None
-    intermediate_reference_points: Optional[torch.FloatTensor] = None
+    last_hidden_state: torch.FloatTensor = None
+    intermediate_hidden_states: torch.FloatTensor = None
+    intermediate_logits: torch.FloatTensor = None
+    intermediate_reference_points: torch.FloatTensor = None
+    intermediate_predicted_corners: Optional[torch.FloatTensor] = None
+    initial_reference_points: Optional[torch.FloatTensor] = None
```
same here

VladOS95-cyber replied:
So, do i have to make all fields optional? because new ones intermediate_predicted_corners and initial_reference_points are already optional

qubvel replied:
Anything that is deafults to `None` should be annotated as Optional

VladOS95-cyber replied:
got it, I'll fix that

qubvel commented on src/transformers/models/rt_detr/modeling_rt_detr.py:
```diff
@@ -292,10 +304,12 @@ class RTDetrObjectDetectionOutput(ModelOutput):
-    intermediate_hidden_states: Optional[torch.FloatTensor] = None
-    intermediate_logits: Optional[torch.FloatTensor] = None
-    intermediate_reference_points: Optional[torch.FloatTensor] = None
+    last_hidden_state: torch.FloatTensor = None
+    intermediate_hidden_states: torch.FloatTensor = None
+    intermediate_logits: torch.FloatTensor = None
+    intermediate_reference_points: torch.FloatTensor = None
+    intermediate_predicted_corners: Optional[torch.FloatTensor] = None
+    initial_reference_points: Optional[torch.FloatTensor] = None
```
same here

qubvel commented on src/transformers/models/rt_detr_v2/modeling_rt_detr_v2.py:
```diff
@@ -549,10 +555,12 @@ class RTDetrV2ModelOutput(ModelOutput):
-    intermediate_hidden_states: Optional[torch.FloatTensor] = None
-    intermediate_logits: Optional[torch.FloatTensor] = None
-    intermediate_reference_points: Optional[torch.FloatTensor] = None
+    last_hidden_state: torch.FloatTensor = None
+    intermediate_hidden_states: torch.FloatTensor = None
+    intermediate_logits: torch.FloatTensor = None
+    intermediate_reference_points: torch.FloatTensor = None
+    intermediate_predicted_corners: Optional[torch.FloatTensor] = None
+    initial_reference_points: Optional[torch.FloatTensor] = None
```
and here

qubvel reviewed (approved):
Thanks for addressing the comments and iterating! Great work, it's ready IMO 🚀

qubvel commented on docs/source/en/index.md:
```diff
@@ -43,3 +43,4 @@ Transformers is designed for developers and machine learning engineers and resea
   </a>
 </div>
 
+Join us on the Hugging Face [Hub](https://huggingface.co/), [Discord](https://discord.com/invite/JfAtkvEtRb), or [forum](https://discuss.huggingface.co/) to collaborate and build models, datasets, and applications together.
```
hmm 🤔

VladOS95-cyber replied:
oops, my bad, fixed

VladOS95-cyber committed:
some posr merge fixes

VladOS95-cyber committed:
small fix

VladOS95-cyber commented on src/transformers/models/rt_detr_v2/modeling_rt_detr_v2.py:
```diff
@@ -641,10 +653,12 @@ class RTDetrV2ObjectDetectionOutput(ModelOutput):
+    intermediate_logits: torch.FloatTensor = None
+    intermediate_reference_points: torch.FloatTensor = None
+    intermediate_predicted_corners: Optional[torch.FloatTensor] = None
+    initial_reference_points: Optional[torch.FloatTensor] = None
```
fixed

VladOS95-cyber committed:
last field fix

VladOS95-cyber commented:
> Thanks for addressing the comments and iterating! Great work, it's ready IMO 🚀

Thanks a lot for review and support!

qubvel commented:
run-slow: d_fine

qubvel commented:
cc @ArthurZucker for final review

The model is based on RT-DETR. The loss implementation is different, so we added some additional attributes to the RT-DETR output to propagate it through the model and avoid overriding all corresponding forward methods. It's a bit against our philosophy, but I believe it is worth it in order to make the code more standard and modular

VladOS95-cyber committed:
fix not split for hgnet_v2

VladOS95-cyber commented:
run-slow: d_fine

qubvel commented:
run-slow: d_fine, hgnet_v2

VladOS95-cyber committed:
disable parallelism test for hgnet_v2 image classification

VladOS95-cyber committed:
skip multi gpu for d-fine

qubvel commented:
run-slow: d_fine, hgnet_v2

qubvel commented:
run-slow: d_fine, hgnet_v2

ArthurZucker commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -0,0 +1,1213 @@
+
+
+class DFineRepNCSPELAN4(nn.Module):
+    # csp-elan
```
if we comment we need a reference to the paper / link to sometyhing

ArthurZucker reviewed (approved):
Happy to merge if you like it @qubvel ! 🤗 
LGTM great review and great work!

VladOS95-cyber committed:
Merge remote-tracking branch 'upstream/main' into add-dfine-model

VladOS95-cyber committed:
adjust after merge init

VladOS95-cyber committed:
remove extra comment

VladOS95-cyber committed:
Merge remote-tracking branch 'upstream/main' into add-dfine-model

VladOS95-cyber committed:
fix repo name references

VladOS95-cyber committed:
small fixes for tests

qubvel committed:
Merge branch 'main' into add-dfine-model

qubvel committed:
Fix checkpoint path

qubvel committed:
Fix consistency

qubvel committed:
Fixing docs

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_36261). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

qubvel commented:
Merging! @VladOS95-cyber thanks a lot for contributing and iterating on the PR 🤗 Really huge work done!

VladOS95-cyber commented:
> Merging! @VladOS95-cyber thanks a lot for contributing and iterating on the PR 🤗 Really huge work done!

Thank you very much for your support and review!

Matvezy commented:
Hello @VladOS95-cyber @qubvel thanks a lot for adding D-FINE to transformers! I'm interested in fine-tuning and benchmarking your implementation on a custom dataset. I found a fine-tuning notebook example [here](https://github.com/qubvel/transformers-notebooks/blob/main/notebooks/DFine_finetune_on_a_custom_dataset.ipynb), however I noticed that it doesn't fully follow the original D-FINE fine-tuning config. I specifically mean not using the same augmentations, optimizer parameters, and not using use EMA like in the original D-FINE configs [here](https://github.com/Peterande/D-FINE/blob/master/configs/dfine/custom/dfine_hgnetv2_l_custom.yml). 
I wanted to ask if you previously tried fine-tuning with the original D-FINE recipe? I suspect that not having it in place may result in a lower benchmark score, so I wanted to ensure I try fine-tuning with the original recipe as well. Right now I'm just adding those changes myself, but I wanted to see if you may already have anything you tired that I can use?

qubvel commented:
Hi @Matvezy! We did not adapt the full fine-tuning recipe for D-FINE. The notebook you mentioned is just a starting point that can be significantly improved with many techniques. We would appreciate any contribution, e.g. if you have some results on your fine-tuning or upgrade the notebook 🤗
